barretenberg/cpp/src/barretenberg/honk/proof_system/permutation_library.hpp

```diff
@@ -166,4 +166,229 @@ void compute_permutation_grand_products(std::shared_ptr<typename Flavor::Proving
+ * maximum value. What can we do in that case? We still have to add (max_range/3)+1 values  to each of the ordered
+ * wires for the sort constraint to hold.  So we also need a and extra denominator to store k ⋅ ( max_range / 3 + 1 )
+ * values that couldn't go in + ( max_range / 3 +  1 ) connecting values. To counteract the extra ( k + 1 ) ⋅
+ * ⋅ (max_range / 3 + 1 ) values needed for denominator sort constraints we need a polynomial in the numerator. So we
+ * can construct a proof when ( k + 1 ) ⋅ ( max_range/ 3 + 1 ) < concatenated size
+ *
+ * @tparam Flavor
+ * @tparam StorageHandle
+ * @param proving_key
+ */
+template <typename Flavor, typename StorageHandle>
+void compute_goblin_translator_range_constraint_ordered_polynomials(StorageHandle* proving_key)
+{
+
+    using FF = typename Flavor::FF;
+
+    // Get constants
+    constexpr auto sort_step = Flavor::SORT_STEP;
+    constexpr auto num_concatenated_wires = Flavor::NUM_CONCATENATED_WIRES;
+    constexpr auto full_circuit_size = Flavor::FULL_CIRCUIT_SIZE;
+    constexpr auto mini_circuit_size = Flavor::MINI_CIRCUIT_SIZE;
+
+    // The value we have to end polynomials with
+    constexpr uint32_t max_value = (1 << Flavor::MICRO_LIMB_BITS) - 1;
+
+    // Number of elements needed to go from 0 to MAX_VALUE with our step
+    constexpr size_t sorted_elements_count = (max_value / sort_step) + 1 + (max_value % sort_step == 0 ? 0 : 1);
+
+    // Check if we can construct these polynomials
+    static_assert((num_concatenated_wires + 1) * sorted_elements_count < full_circuit_size);
+
+    // First use integers (easier to sort)
+    std::vector<size_t> sorted_elements(sorted_elements_count);
+
+    // Fill with necessary steps
+    sorted_elements[0] = max_value;
+    for (size_t i = 1; i < sorted_elements_count; i++) {
+        sorted_elements[i] = (sorted_elements_count - 1 - i) * sort_step;
+    }
+
+    std::vector<std::vector<uint32_t>> ordered_vectors_uint(num_concatenated_wires);
+    auto ordered_constraint_polynomials = std::vector{ &proving_key->ordered_range_constraints_0,
+                                                       &proving_key->ordered_range_constraints_1,
+                                                       &proving_key->ordered_range_constraints_2,
+                                                       &proving_key->ordered_range_constraints_3 };
+    std::vector<size_t> extra_denominator_uint(full_circuit_size);
+
+    // Get information which polynomials need to be concatenated
+    auto concatenation_groups = proving_key->get_concatenation_groups();
+
+    // A function that transfers elements from each of the polynomials in the chosen concatenation group in the uint
+    // ordered polynomials
+    auto ordering_function = [&](size_t i) {
+        // Get the group and the main target vector
+        auto my_group = concatenation_groups[i];
+        auto& current_vector = ordered_vectors_uint[i];
+        current_vector.resize(Flavor::FULL_CIRCUIT_SIZE);
```

Comment:
can we add a TODO here for a future optimisation?

`resize` is an expensive function because it performs a deep copy of the existing vector elements into new memory.

It would be more efficient to construct these vectors with the correct size before we populate them.

Reply:
Actually, resize is not needed here, since we are not using the optimization yet

Reply:
These vectors are empty initially. I resized because there was a vector of vectors. Internal vectors are empty